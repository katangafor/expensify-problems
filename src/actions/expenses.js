import database from '../firebase/firebase';

// ADD_EXPENSE
// this one parameter is an object, which has default parameters if fed an object, and defaults to an empty object if fed no object
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
})

export const startAddExpense = (expenseData = {}) => {
  // this part only works cause of redux-thunk middleware
  return (dispatch) => {
    const {
      description = '', 
      note = '', 
      amount = 0, 
      createdAt = 0 
    } = expenseData;
    const expense = { description, note, amount, createdAt };

    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }))
    })
  }
}

// REMOVE EXPENSE
export const removeExpense = ({ id }) => ({
  type: 'REMOVE_EXPENSE',
  id
})

// START REMOVE EXPENSES
export const startRemoveExpense = ({ id }) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({ id }))
    })
  }
}

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})

// SET EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

// START SET EXPENSES
export const startSetExpenses = (expenses) => {
  return (dispatch) => {
    return database.ref('expenses').once('value').then((snapshot) => {
      const expenses = [];
      snapshot.forEach((childSnapshot) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        })
      })
      dispatch(setExpenses(expenses))
    })
  }
}