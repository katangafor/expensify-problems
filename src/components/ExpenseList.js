import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

// this makes a named export out of this component, so I can use it in tests without messing with the default export for redux stuff
export const ExpenseList = (props) => (
  <div>
  {
    props.expenses.length === 0 ? (
       <p>No expenses</p>
    ) : (
      props.expenses.map((expense) => {
        // I thought the spread operator just returns a new object? So why here does it just work to populate three props? Can you
        // just pass in an object and have it become props?
        return <ExpenseListItem key={expense.id} {...expense}/>
      })
    )
  }
  </div>
)

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  }
}

export default connect(mapStateToProps)(ExpenseList);

