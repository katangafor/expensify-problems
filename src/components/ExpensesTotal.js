import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesTotal = (props) => (
  <div>
    {
      props.expenses.length === 0 ? (
        <p>Dam u lucky</p>
      ) : (
        <p>Found {props.expenses.length} expenses totaling {numeral(getExpensesTotal(props.expenses) / 100).format('$0,0.00')}</p>
      )
    }
  </div>
)

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  }
}

export default connect(mapStateToProps)(ExpensesTotal);