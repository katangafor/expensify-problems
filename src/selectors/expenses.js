

// Takes in the filters and the sort criteria, and reeturns a new array that's properly sorted
export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter((expense) => {
    // this is probably better done with momment() methods
    const startDateMatch = startDate !== undefined ? expense.createdAt >= startDate : true;
    const endDateMatch = endDate !== undefined ? expense.createdAt <= endDate : true;
    const textMatch = text ? expense.description.toLowerCase().includes(text.toLowerCase()) : true;

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1;
    }
  });
}