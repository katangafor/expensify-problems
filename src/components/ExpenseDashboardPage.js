import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesTotal from './ExpensesTotal';

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpensesTotal />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;