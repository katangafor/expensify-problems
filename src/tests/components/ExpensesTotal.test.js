import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesTotal } from '../../components/ExpensesTotal';
import expenses from '../fixtures/expenses';

test('should render expenses total correctly', () => {
  const wrapper = shallow(<ExpensesTotal expenses={expenses} />)
  expect(wrapper).toMatchSnapshot();
})

test('should render text if no expenses', () => {
  const wrapper = shallow(<ExpensesTotal expenses={[]} />)
  expect(wrapper).toMatchSnapshot();
})