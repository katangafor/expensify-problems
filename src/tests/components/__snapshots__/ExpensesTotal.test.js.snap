// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expenses total correctly 1`] = `
<div>
  <p>
    Found 
    3
     expenses totaling 
    $6.00
  </p>
</div>
`;

exports[`should render text if no expenses 1`] = `
<div>
  <p>
    Dam u lucky
  </p>
</div>
`;
